core/src/main/java/org/apache/iceberg/IncrementalDataScan.java

```diff
@@ -0,0 +1,119 @@
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Predicate;
+import com.google.common.collect.Sets;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.Set;
+import org.apache.iceberg.expressions.Expression;
+import org.apache.iceberg.io.CloseableIterable;
+
+class IncrementalDataScan extends DataTableScan implements TableScan {
+  private final TableOperations ops;
+  private final Table table;
+  private boolean colStats;
+  private long fromSnapshotId;
+  private long toSnapshotId;
+  private Set<Long> snapshotIds;
+
+
+  IncrementalDataScan(TableOperations ops, Table table, long fromSnapshotId, long toSnapshotId) {
+    super(ops, table);
+    this.ops = ops;
+    this.table = table;
+    this.fromSnapshotId = fromSnapshotId;
+    this.toSnapshotId = toSnapshotId;
+    this.snapshotIds = snapshotsWithin(table, fromSnapshotId, toSnapshotId);
+  }
+
+  /**
+   * @return all snapshots ids between {@param fromSnapshotId } exclusive
+   * and {@param toSnapshotId} inclusive
+   */
+  private static Set<Long> snapshotsWithin(Table table, long fromSnapshotId, long toSnapshotId) {
+    Set<Long> snapshotIds = Sets.newHashSet();
+    // table history is ordered
+    Iterator<HistoryEntry> historyEntryIterator = table.history().iterator();
+    while (historyEntryIterator.hasNext()) {
+      if (historyEntryIterator.next().snapshotId() == fromSnapshotId) {
+        break;
+      }
+    }
+    while (historyEntryIterator.hasNext()) {
+      long snapshotId = historyEntryIterator.next().snapshotId();
+      snapshotIds.add(snapshotId);
+      if (snapshotId == toSnapshotId) {
+        break;
+      }
+    }
+    return snapshotIds;
+  }
+
+  private IncrementalDataScan(TableOperations ops, Table table, Long snapshotId, Schema schema,
+                              Expression rowFilter, boolean caseSensitive, boolean colStats,
+                              Collection<String> selectedColumns,
+                              Long fromSnapshotId, Long toSnapshotId) {
+    super(ops, table, snapshotId, schema, rowFilter, caseSensitive, colStats, selectedColumns);
+    this.ops = ops;
+    this.table = table;
+    this.colStats = colStats;
+    this.fromSnapshotId = fromSnapshotId;
+    this.toSnapshotId = toSnapshotId;
+  }
+
+
+  @Override
+  public CloseableIterable<FileScanTask> planFiles() {
+    Snapshot snapshot = table.snapshot(toSnapshotId);
+    Preconditions.checkState(snapshot != null, "Cannot find snapshot corresponding to snapshot id: %s", toSnapshotId);
+    //TODO publish an incremental scan event
+    return planFiles(ops, snapshot, filter(), isCaseSensitive(), colStats);
+  }
+
+  @Override
+  @SuppressWarnings("checkstyle:HiddenField")
+  public CloseableIterable<FileScanTask> planFiles(
+          TableOperations ops, Snapshot snapshot, Expression rowFilter, boolean caseSensitive, boolean colStats) {
```

Comment:
I think it might be a bit cleaner to implement `planFiles` using `Snapshot.addedFiles` since this has to add methods to pass in predicates for matching manifests and manifest entries. The advantage of `addedFiles` is that you don't need to scan compacted manifests from replace commits. Ignoring those large manifests is probably faster in most cases, especially if your writer is using `FastAppend`.

Reply:
Hi @rdblue I'm looking into it again. This sounds good, though one open question for me is how do we get the right schema to pass into a `MetricsEvaluator` which will be required to filter data-files.  With the current approach, using manifests, we get the right schema from the `partitionSpec` stored in the manifest.

Reply:
@rdblue  I couldn't directly use `Snapshot.addedFiles` ,  but  I've incorporated your idea in the updated PR which allows me the pass in the right parameters when reading manifests. With the updated approach, I get the best of both worlds, e.g I can now ignore manifests and snapshots from replace commits